### ISSUE TITLE
Deprecate MPI_PROD

### ISSUE THREAD

**Issue by Wee-Free-Scot** [mpi-forum/mpi-issues#567]
# Problem

There are no known usages of MPI_PROD in application codes. Implementing MPI_PROD in RMA efficiently requires hardware that does not exist; being forced to implement a slow-path MPI_PROD for RMA compromises the fast-path implementation of commonly-used operators in RMA. Deprecating for RMA but not for reductions and scans would suggest there is a use-case for MPI_PROD in collectives, but there is no such use-case.

# Proposal

Deprecate MPI_PROD throughout the MPI Standard, with a view to removing it in future.

# Changes to the Text

Remove MPI_PROD from the table of predefined reduction operators (page 215&216, section 6.9.2).
Remove MPI_PROD from the table of collective operations [sic] (page 824, Appendix A).
Add a new section into the Deprecated Interfaces chapter recording the MPI_PROD constant (page 750, section 16.4).

# Impact on Implementations

Joy and celebration.

# Impact on Users

Meh.

# References and Pull Requests

https://github.com/mpi-forum/mpi-standard/pull/703

**Comment by Wee-Free-Scot** [comment]
Here is a PDF with the proposed changes:
[mpi40-report_Issue567_Deprecate_MPI_PROD.pdf](https://github.com/mpi-forum/mpi-issues/files/9201434/mpi40-report_Issue567_Deprecate_MPI_PROD.pdf)

Removals on pages 215, 216, and 824, addition on page 750, adjustment to constant index on page 1054.

**Comment by wesbland** [comment]
![image](https://user-images.githubusercontent.com/1561207/182634399-11c2566f-7bc0-4235-93d6-6a1707248c9b.png)

**Comment by devreal** [comment]
@RolfRabenseifner dug up this code snippet which seems legit (i.e., not just a wrapper or test case): https://github.com/manolillo/palm4u/blob/1751eca6b334bb29c23f85fdb1d78f22da34eb7a/current_version/trunk/SOURCE/spectra_mod.f90#L690

**Comment by christgau** [comment]
And It's still present in the latest (original) code: https://gitlab.palm-model.org/releases/palm_model_system/-/blob/master/packages/palm/model/src/spectra_mod.f90#L685

**Comment by jeffhammond** [comment]
PVM was used too. Now it's not. People will find a way. If somebody wants to use this ridiculous feature, that can write a user defined reduction. I'll do it for them if they complain.

**Comment by jeffhammond** [comment]
Also, why the code comment say "Global sum of spectra on PE0" if it's doing PROD? Does this code actually _work_?

**Comment by jeffhammond** [comment]
Also, __parallel is a reserved symbol per the C standard so this code is UB.

**Comment by Wee-Free-Scot** [comment]
Do we wish to:
1. carry on regardless -- ignore the potential for existing usages in the wild and deprecate MPI_PROD entirely in MPI-4.1?
2. roll with it -- grudgingly accept the possibility of MPI_REDUCE usage and restrict the deprecation to RMA accumulate operations in MPI-4.1?
3. give up (for now) -- either close this and never speak of it again or mark it for MPI-5.0 and hope for the best?

I vote for option (2) -- that addresses the performance critical issue without side-effects elsewhere.

**Comment by cniethammer** [comment]
I'd be in favor of option (1) -- keeping it simple and not needing an(other) exception for the usage of reduction operations in Collective/reduce vs RMA/accumulate operations.
However, I would also support option (2) to improve at least the RMA part.
I think this should be addressed in 4.1 to go forward.

For those really needing a global product I would add an advice to users:
* warning about overflows at scale
* pointing out the math solution using the equation
  ![image](https://user-images.githubusercontent.com/8720191/197175453-40a21ac5-1946-4745-8e46-ded0acb4eaf2.png)
  usable with `MPI_SUM` for reduction and accumulate operations.

**Comment by VictorEijkhout** [comment]
> On , 2022Oct21, at 05:56, Christoph Niethammer ***@***.***> wrote:
> 
> out the math solution

For positive numbers.

V.

**Comment by Wee-Free-Scot** [comment]
> For positive numbers.

Correct, you'd need a second element in the reduction that counted negative inputs (each process supplies 0 for positive or 1 for negative -- that is, give `{ log(abs(a_i)), (a_i<0)?1:0 }` as input). If the count is odd, negate the result of the exponential.

**Comment by Wee-Free-Scot** [comment]
Superseded by #559

**Comment by jeffhammond** [comment]
I think we need do so PROD without M**LOC based on the discussion on the other ticket.

**Comment by Wee-Free-Scot** [comment]
> I think we need do so PROD without M**LOC based on the discussion on the other ticket.

Which other issue?

**Comment by jeffhammond** [comment]
https://github.com/mpi-forum/mpi-issues/issues/559#issuecomment-1191666793

**Comment by Wee-Free-Scot** [comment]
> I think we need do so PROD without M**LOC based on the discussion on the other ticket.

Sorry, I'm struggling to parse this collection of words.

We must now retarget for MPI-4.2 or MPI-5.0, whichever happens first (and permits non-ABI changes).

Do we wish to:

1. carry on regardless -- ignore the potential for existing usages in the wild and deprecate MPI_PROD entirely in MPI-4.2/MPI-5.0 (if so, advise users to use UDF when they simply must have product)?
2. roll with it -- grudgingly accept the possibility of MPI_REDUCE usage and restrict the deprecation to RMA accumulate operations in MPI-4.2/MPI-5.0 (if so, see issue #559)?
3. give up -- close this and never speak of it again?

I prefer option (1) -- that is my personal preference.

I vote for option (2) -- that addresses the performance critical issue without side-effects elsewhere.

I implemented option (3) -- that is the one most likely to gain broad acceptance in the MPI Forum.

**Comment by jeffhammond** [comment]
I want to deprecate `MPI_PROD` in MPI 5.0.  If you don't want to work on it, I will.

**Comment by Wee-Free-Scot** [comment]
I did work on it -- hence this issue and the associated PR.

AFAIC, it's ready to go; it just needs sufficient support from MPI Forum members to pass a couple of votes.

If I missed somewhere in the document where text needs to be removed/changed/added, then let me know and I'll take look.

Like you @jeffhammond, I am frustrated that this clear improvement to MPI seems to be stymied by terrible examples of legacy usage.

**Comment by jeffhammond** [comment]
If we can deprecate MPIF.H, we can deprecate MPI_PROD.

**Comment by jeffhammond** [comment]
We should start reading this.  I am sorry we missed it for MPI-5 but it's my highest priority for MPI-Next.

**Comment by wesbland** [comment]
Are you going to be at the next meeting? Want to announce it?

**Comment by jeffhammond** [comment]
I won't be in Vienna until Thursday. I can only do a reading on Monday.

**Comment by wesbland** [comment]
Monday is fine. You're saying you want to do it remotely?
